apis/pkg/v1alpha1/controllerrevision_types.go

```diff
@@ -0,0 +1,264 @@
+// Copyright 2024 Upbound Inc.
+// All rights reserved
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	pkgv1 "github.com/crossplane/crossplane/apis/pkg/v1"
+)
+
+// ControllerPackagingType defines an enum for the controller package type.
+type ControllerPackagingType string
+
+const (
+	// ControllerPackagingTypeHelm represents a controller that is packaged as a
+	// Helm chart.
+	ControllerPackagingTypeHelm ControllerPackagingType = "Helm"
+)
+
+// +kubebuilder:object:root=true
+// +genclient
+// +genclient:nonNamespaced
+
+// An ControllerRevision represents a revision of an Controller.
+//
+// +kubebuilder:subresource:status
+// +kubebuilder:storageversion
+// +kubebuilder:printcolumn:name="HEALTHY",type="string",JSONPath=".status.conditions[?(@.type=='Healthy')].status"
+// +kubebuilder:printcolumn:name="REVISION",type="string",JSONPath=".spec.revision"
+// +kubebuilder:printcolumn:name="IMAGE",type="string",JSONPath=".spec.image"
+// +kubebuilder:printcolumn:name="STATE",type="string",JSONPath=".spec.desiredState"
+// +kubebuilder:printcolumn:name="DEP-FOUND",type="string",JSONPath=".status.foundDependencies"
+// +kubebuilder:printcolumn:name="DEP-INSTALLED",type="string",JSONPath=".status.installedDependencies"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:resource:scope=Cluster,categories={upbound,pkgrev}
+type ControllerRevision struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ControllerRevisionSpec   `json:"spec"`
```

Comment:
nit: to make `spec.helm` immutable at validation time:
```suggestion
	// +kubebuilder:validation:XValidation:rule="has(self.helm) == has(oldSelf.helm)",message="spec.helm specification is immutable"
	Spec   ControllerRevisionSpec   `json:"spec"`
```

Reply:
Hmm, this didn't really work

```
message: 'cannot persist helm spec: cannot update package revision spec: ControllerRevision.pkg.upbound.io
      "kyverno-e28cf75aea8d" is invalid: spec: Invalid value: "object": spec.helm
      specification is immutable'
```

Reply:
Ok, it looks like the following is working fine:

```
// +kubebuilder:validation:XValidation:rule="!has(oldSelf.helm) || self.helm == oldSelf.helm",message="helm specification is immutable"
```